index.html

```diff
@@ -0,0 +1,344 @@
+        container.append(header);
+
+        loader.start();
+
+        const data = await fetchRepos({
+          url,
+          cache: createCache(`${sort}-${limit}`),
+        });
+
+        if (data.error) {
+          loader.stop("✗");
+
+          container.classList.add("error");
+          container.append(data.message);
+        } else {
+          loader.stop("✔");
+
+          if (data.cacheHit) {
+            const cacheStatus = createCacheStatus(data.expirationTime);
+            header.append(cacheStatus);
+          } else {
+            const rateLimit = createRateLimit(data.rateLimit);
+            header.append(...rateLimit);
+          }
+
+          const body = createBody();
+          const repos = createRepos({
+            container: body,
+            repos: data.repos,
+            metaFields,
+          });
+          body.append(...repos);
+          container.append(body);
+        }
+      }
+
+      function createRateLimit(rateLimit) {
+        return rateLimit.map((info) => {
+          const p = document.createElement("p");
+          p.textContent = info.join(": ");
+          return p;
+        });
+      }
+
+      function createHeader(url, loader) {
+        const header = document.createElement("header");
+        const title = document.createElement("h2");
+        const link = document.createElement("a");
+        link.href = url;
+        link.textContent = url;
+        title.append(loader.ref, "Fetching ", link);
+        header.append(title);
+
+        return header;
+      }
+
+      function createBody() {
+        const projectsBody = document.createElement("div");
+        projectsBody.classList.add("projects-body");
+
+        return projectsBody;
+      }
+
+      function createCacheStatus(expirationTime) {
+        const container = document.createElement("div");
+        container.innerHTML = `
+          <p>Cache hit!</p>
+          <p>Expires: ${expirationTime}</p>
+        `;
+        return container;
+      }
+
+      function createLoader() {
+        let interval = null;
+        const element = document.createElement("span");
+        element.classList.add("repo-loader");
+
+        return {
+          ref: element,
+          start: () => {
+            const loaderSymbols = ["\\", "|", "/", "—"];
+            let i = 0;
+            element.textContent = loaderSymbols[i % loaderSymbols.length];
+            interval = setInterval(() => {
+              console.log("ticking");
+              element.textContent = loaderSymbols[++i % loaderSymbols.length];
+            }, 500);
+          },
+          stop: (textContent = "") => {
+            clearInterval(interval);
+            element.textContent = textContent;
+          },
+        };
+      }
+
+      function createCache(key) {
+        return {
+          get() {
+            const cache = JSON.parse(localStorage.getItem(key));
+            if (cache?.expirationTime > Date.now()) {
```

Comment:
Why using the existential operator here? As far as I can see, the `expirationTime` property would be defined always.
```suggestion
            if (cache.expirationTime > Date.now()) {
```

Reply:
Great observation! The issue is that `cache` is null if you open the page for the first time, and trying to access a property on null throws an error (`cannot read property 'expirationTime' of null). That's why it is needed.